### PR TITLE
Basic filtering for "nsfw" keyword

### PR DESCRIPTION
**Pull request by bmarty**
New settings, in Settings/Preferences screen, off by default:

(Edit: wording has been updated, see new screenshot below)
image

Filtering room directory

image

Also migrate RoomDirectoryService to coroutines (#2449)

### CODE REVIEW

**Comment by SimonBrandner** [comment]
May I have two questions here? 
+ I _think_ "Filter NSFW rooms" might be easier to find. 
+ How exactly does the filtering work? The Blender room, for example, has _No NSFW Media_ in its name. Would it be hidden then?

**Comment by bmarty** [comment]
>     * I _think_ _Filter NSFW rooms_ might be easier to find.

"NSFW" is not understandable enough for non power users.

>     * How exactly does the filtering work? The Blender room, for example, has _No NSFW Media_ in its name. Would it be hidden then?

Yes, I have seen the problem with the Blender room, it will be actually filtered out. Filtering like that is far from ideal. This is a first quick implementation, we will do better in the future (filtering server side, new state event describing the room content, etc.)

**Comment by SimonBrandner** [comment]
Thanks for the quick response!

> "NSFW" is not understandable enough for non power users.

Maybe _"Filter explicit content"_ or something like that. I would just expect the _explicit_ word to be in the "header".

But I am probably just nitpicking at this point... :joy:  Sorry for annoying

**Comment by bmarty** [comment]
I have created https://github.com/matrix-org/matrix-doc/issues/3054 to try to do better on this subject

**Comment by bmarty** [comment]
Update wording

image

**Comment by bmarty** [comment]
FTR I have asked the admin of the room about Blender to update the topic of the room to avoid being filtered by mistake.

File: vector/src/main/java/im/vector/app/features/roomdirectory/RoomDirectoryViewModel.kt [hunk 1]
```diff
@@ -155,47 +166,60 @@ class RoomDirectoryViewModel @AssistedInject constructor(@Assisted initialState:
...
+
+            currentJob = null
+
+            since = data.nextBatch
+
+            // Filter
+            val newPublicRooms = data.chunk.orEmpty()
+                    .filter {
+                        showAllRooms
+                                || "${it.name.orEmpty()} ${it.topic.orEmpty()} ${it.canonicalAlias.orEmpty()}".toLowerCase(Locale.ROOT)
```
Comment (onurays) [inline vector/src/main/java/im/vector/app/features/roomdirectory/RoomDirectoryViewModel.kt:206]: I couldn't suggest a better way for now, will think later.

**Pull request merged**